### ISSUE TITLE
Unintuitive condition for labelProps 

### ISSUE THREAD

**Issue by Andarist** [adobe/react-spectrum#1476]
# 📝 Feedback

`labelProps` generated by `useLabel` that is used by hooks like `useTextField` are quite convenient:
https://github.com/adobe/react-spectrum/blob/7927421dedfb001d3459e0a94f5def9461fbc4c7/packages/%40react-aria/label/src/useLabel.ts#L50-L54

However, I find the condition based on which those are returned quite not intuitive. Since a hook like `useTextField` is not directly responsible for managing the label (which manifests in the label not being returned from it) it's odd to even pass a `label` prop to it because it seems unrelated. 

I would mostly expect hooks to accept props that are needed for managing internal state, inter-hook communication, or which translate to returned props that are supposed to be later spread onto the DOM. This one seems like a too indirect translation of the `label` prop to `{ labelProps: { id, htmlFor } }`.

## 🔦 Context

I've struggled to see what's the preferred way of using this hook in combination with the existing, static, label.

## 💻 Code Sample

Consider a simple static label example:
```js
function Foo(props) {
  const inputRef = React.useRef(null)
  const { inputProps, labelProps } = useTextField(props, inputRef)

  return (
    <>
      <label {...labelProps}>Some label</label>
      <input {...inputProps}/>
    </>
  )
}
```

I think it's far from obvious/intuitive to know that this could be refactored to:
```js
function Foo(props) {
  const inputRef = React.useRef(null)
  const label = 'Some label'
  const { inputProps, labelProps } = useTextField({ ...props, label }, inputRef)

  return (
    <>
      <label {...labelProps}>{label}</label>
      <input {...inputProps}/>
    </>
  )
}
```
Which also breaks the colocation of the label element and label text - for no good reason.

**Comment by ktabors** [comment]
The hooks for inputs, like text fields, call the label hook to make sure there is a label, i.e. there is a warning when there isn't a label. This association is to provide input field context and accessibility functionality. Do you have a suggestion on how to improve the documentation to make this clear?

**Comment by Andarist** [comment]
Well, I understand why this exists there but I believe this, partially, makes for a quirky API. It is convenient for those `labelProps` to be generated by inputs-related hooks **in some scenarios** but maybe in the future, you should consider decoupling those two a little bit more. The validation portion is very helpful, but maybe it should not happen during a render phase but rather in an effect or smth? You could then really validate if something has not been misused and notify the user about the need to correct themselves.

If you plan on not decoupling those 2 - which also would make sense, maybe you could consider input-related hooks to **accept** `labelProps`. This would enforce good composition traits at expense of the developer having to call `useLabel` themselves. 

As to the docs - just a note explaining the relation between both and how things are supposed to be used in different scenarios would make a huge difference. I certainly don't want to bitch about anything here because this project solves a lot of problems that we've been dealing with in-house before starting to use React Aria so we very much appreciate this! But to give some feedback from my side - when using the project I often struggle to compose things correctly from the start. I need to dig into the implementation to see how things relate to each other and I often check how hooks are utilized by your component library - so I feel that some information is lacking from the docs but I also understand that documenting things is just very hard and this is not your top priority. It's already nice-enough of you that this project has been open-sourced. A second feeling I often have is that some things have really been not designed in isolation and they assume the certain structure of the consuming components and their props - it feels a little bit that the Spectrum API has influenced greatly some API choices of the React Aria. Which again - it's your project, it makes sense from your perspective if you don't want to overabstract and the extracted bits are still usable on their own so that's great. Using those bits with different component APIs is a little bit quirky at times though.

**Comment by ktabors** [comment]
That is an interesting suggestion on decoupling these in this way for future versions. Thanks! Like you mentioned we are open-source and appreciate contributions. If you have the bandwidth to improve these docs that would be helpful and appreciated since we're currently writing docs for some other components and adding more components to the project.

**Comment by Andarist** [comment]
As an open-source maintainer myself I know how valuable docs PRs are. However, being a non-native speaker this is somewhat harder for me to just do quickly and my OSS bandwidth is already too packed 😢 I have been providing some PRs to React Aria in the past though and I plan to continue to do so in the future - I focus on code/functionality improvements though.

**Comment by ktabors** [comment]
Not making a docs PR sense and thank you for your contributions. We'll keep the docs improvement in mind.

**Comment by tuananhlai** [comment]
I agree with the OP that the API for labeling is not quite intuitive. If I want to use an external label for my `RadioGroup` component, I'd have to provide a workaround to suppress the warning below.

```
If you do not provide a visible label, you must specify an aria-label or aria-labelledby attribute for accessibility
```

**Comment by snowystinger** [comment]
You shouldn't need to provide a "workaround", just pass in aria-labelledby pointing to your external label. That is the purpose of this warning, to make sure that labels are wired up correctly.

**Comment by tuananhlai** [comment]
@snowystinger Thank you for your answer. I understand that for complex component like combo box, using `aria-labelledby` to point to the external label is necessary. But for simple component like text field, is `aria-labelledby` necessary? I think an external label with the correct `for` attribute would suffice in that case, so I find

**Comment by snowystinger** [comment]
You can also pass it the label and then spread the results of `useLabel` onto that label. This will ensure that you have the correct `htmlFor` and an id, there is no chance of misspelling the id or forgetting that in react it is `htmlFor` not `for`.

It's a little quirky, as noted up above, and we'd love some docs to clarify it. We will also consider proposed changes, especially if they are non-breaking.

**Comment by tuananhlai** [comment]
@snowystinger While I find `useLabel` useful in solving some problems, sometime an external label is better for more flexibility in styling and positioning the label. In those cases, the warning is a bit counter-productive. It'd be great if we have an option to suppress the warning when using external label. 💭
